### PR TITLE
1998  User Schedule Set Endpoint

### LINKED ISSUE #1998: Design Review Calendar - Set User Schedule Settings Endpoint
### Description

Blocked by: #1846

We need a way for users to set their schedule settings to schedule design reviews. This ticket is to create a backend endpoint that sets (this uses the upsert operation) a user's schedule settings.

### Acceptance Criteria

- [x] Add an endpoint to set schedule settings
- [x] It should be on the route '/users/schedule-settings/set'
- [x] Payload should include:
       - personalGmail (String)
       - personalZoomLink (String)
       - availability (int[])
- [x] The settings should be set for the current user
- [x] Add tests
- [x] Manually test using Postman, and include screenshots of database on prisma studio before and after calling endpoint (reach out to Sean if you have questions)

### Proposed Solution

See setUserSecureSettings for an example of how to use upsert

### Mocks

_No response_

### PR DESCRIPTION
**Pull request by caiodasilva2005**
## Changes

Added endpoint for setting user schedule

## Test Cases

Made test to ensure request is returning an id.

image

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [X] All commits are tagged with the ticket number
- [X] No linting errors / newline at end of file warnings
- [X] All code follows repository-configured prettier formatting
- [X] No merge conflicts
- [X] All checks passing
- [X] Remove any non-applicable sections of this template
- [X] Assign the PR to yourself
- [X] No `yarn.lock` changes (unless dependencies have changed)
- [X] Request reviewers & ping on Slack
- [X] PR is linked to the ticket (fill in the closes line below)

Closes #1998

### CODE REVIEW

File: src/backend/src/services/users.services.ts [hunk 1]
```diff
@@ -339,4 +339,27 @@ export default class UsersService {
 
     return userScheduleSettingsTransformer(scheduleSettings);
   }
+
+  static async setUserScheduleSettings(
```
Comment (jackturner106) [inline src/backend/src/services/users.services.ts:343]: Can you check that emails are unique? See how its done for phone numbers in setUserSecureSettings.

File: src/backend/tests/users.test.ts [hunk 2]
```diff
@@ -167,5 +167,13 @@ describe('Users', () => {
       expect(prisma.schedule_Settings.findUnique).toHaveBeenCalledTimes(1);
       expect(res).toStrictEqual(batmanUserScheduleSettings);
     });
+
+    test('setUserScheduleSettings works successfully', async () => {
```
Comment (jackturner106) [inline src/backend/tests/users.test.ts:171]: Could you add tests for invalid emails/invalid zoom links, and maybe also for non-unique emails?

**Review by jackturner106 — CHANGES_REQUESTED** [review]
This is looking good! There's a couple more things I would like to verify on the endpoint. Also, it looks like .[0]x was pushed? Its the file that git uses to track commits, and normally it should not show up, so I'm not actually sure whats happened there. See if you can get it to not show up in the commit, you may need to more or less manually revert it.

File: src/backend/src/controllers/users.controllers.ts [hunk 3]
```diff
@@ -146,6 +146,19 @@ export default class UsersController {
...
   }
 
+  static async setUserScheduleSettings(req: Request, res: Response, next: NextFunction) {
+    try {
+      const { personalGmail, personalZoomLink, availability } = req.body;
+      const user = await getCurrentUser(res);
+
+      await UsersService.setUserScheduleSettings(user, personalGmail, personalZoomLink, availability);
+
+      res.status(200).json({ message: `Successfully updated schedule settings for user ${user.userId}.` });
```
Comment (RChandler234) [inline src/backend/src/controllers/users.controllers.ts:156]: should be returning the updated schedule settings

File: src/backend/src/services/users.services.ts (hunk 1 above)
Comment (RChandler234) [inline src/backend/src/services/users.services.ts:343]: this should have a detailed comment documented inputs/ outputs

**Review by RChandler234 — CHANGES_REQUESTED** [review]

File: src/backend/src/controllers/users.controllers.ts [hunk 4]
```diff
@@ -146,6 +146,24 @@ export default class UsersController {
     }
   }
 
+  static async setUserScheduleSettings(req: Request, res: Response, next: NextFunction) {
+    try {
+      const { personalGmail, personalZoomLink, availability } = req.body;
+      const user = await getCurrentUser(res);
+
+      const editedScheduleSettings = await UsersService.setUserScheduleSettings(
```
Comment (RChandler234) [inline src/backend/src/controllers/users.controllers.ts:154]: *updatedScheduleSettings

File: src/backend/src/services/users.services.ts [hunk 5]
```diff
@@ -339,4 +339,43 @@ export default class UsersService {
...
+  ): Promise<string> {
+    const existingUser = await prisma.schedule_Settings.findFirst({
+      where: { personalGmail, userId: { not: user.userId } } // excludes the current user from check
+    });
+
+    if (existingUser) {
+      throw new HttpException(400, 'Email already in use');
+    }
+
+    const newUserScheduleSettings = await prisma.schedule_Settings.upsert({
```
Comment (RChandler234) [inline src/backend/src/services/users.services.ts:365]: updateScheduleSettings*

File: src/backend/src/services/users.services.ts [hunk 6]
```diff
@@ -339,4 +339,43 @@ export default class UsersService {
...
+        availability
+      },
+      create: {
+        userId: user.userId,
+        personalGmail,
+        personalZoomLink,
+        availability
+      }
+    });
+    return newUserScheduleSettings.drScheduleSettingsId;
```
Comment (RChandler234) [inline src/backend/src/services/users.services.ts:379]: this should return the whole updateScheduleSettings, not just the id

**Review by RChandler234 — CHANGES_REQUESTED** [review]
remove that extra file pls

**Review by RChandler234 — APPROVED** [review]
LGTM 🔥

**Pull request merged**